File: leetcode/medium/RangeSumQuery2D.py

from typing import List

# Time: O(m*n) for __init__, O(1) for sumRegion
# Space: O(m*n) for the prefix sum matrix
class NumMatrix:
    """
    This class allows for fast calculation of the sum of elements within a 2D rectangular region.
    It uses a 2D prefix sum matrix (also known as a summed-area table) to achieve this.
    After an initial O(m*n) pre-computation, region sums can be calculated in O(1) time.
    """

    def __init__(self, matrix: List[List[int]]):
        """
        Initializes the data structure by building a 2D prefix sum matrix.

        :param matrix: A 2D list of integers (the input matrix).
        """
        if not matrix or not matrix[0]:
            # Handle empty or invalid input matrix
            self.prefix = None
            return

        rows, cols = len(matrix), len(matrix[0])
        
        # Initialize a prefix sum matrix with an extra row and column of zeros.
        # This padding simplifies the calculation logic in both __init__ and sumRegion.
        self.prefix = [[0] * (cols + 1) for _ in range(rows + 1)]

        # Build the 2D prefix sum matrix.
        for r in range(rows):
            for c in range(cols):
                # The value at prefix[r+1][c+1] is the sum of the rectangle
                # from (0,0) to (r,c) in the original matrix.
                # Formula: Sum(ABCD) = Sum(D) - Sum(B) - Sum(C) + Sum(A)
                # where D is the current rectangle, B and C are adjacent rectangles,
                # and A is the top-left overlapping rectangle.
                # Rearranging gives: Sum(D) = Val(D) + Sum(B) + Sum(C) - Sum(A)
                self.prefix[r + 1][c + 1] = matrix[r][c] + self.prefix[r][c + 1] + self.prefix[r + 1][c] - self.prefix[r][c]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        """
        Calculates the sum of the elements in the rectangle defined by the top-left
        corner (row1, col1) and the bottom-right corner (row2, col2).

        :param row1: The starting row index of the region.
        :param col1: The starting column index of the region.
        :param row2: The ending row index of the region.
        :param col2: The ending column index of the region.
        :return: The sum of the elements in the specified region.
        """
        if self.prefix is None:
            return 0

        # The sum of a region is calculated using the pre-computed prefix sums.
        # Let the desired region be a rectangle with corners at (row1, col1) and (row2, col2).
        # The sum is: (sum up to bottom-right) - (sum up to area above) - (sum up to area to the left) + (sum of overlapping top-left area).
        # This corresponds to: prefix[row2+1][col2+1] - prefix[row1][col2+1] - prefix[row2+1][col1] + prefix[row1][col1].
        
        # All coordinates are increased by 1 due to the padding in the prefix matrix.
        bottom_right = self.prefix[row2 + 1][col2 + 1]
        above = self.prefix[row1][col2 + 1]
        left = self.prefix[row2 + 1][col1]
        top_left = self.prefix[row1][col1]
        
        return bottom_right - above - left + top_left


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)

File: leetcode/medium/RangeSumQuery2D.java
/**
 * A utility class to perform efficient range sum queries on a 2D immutable matrix.
 * Uses the 2D Prefix Sum algorithm (Inclusion-Exclusion Principle).
 */
public class NumMatrix {
    
    // Using a 1-based index approach for the prefix array to simplify boundary checks.
    // prefixSum[i][j] stores the sum of the sub-matrix from (0,0) to (i-1, j-1).
    private final int[][] prefixSum;
    private final boolean isEmpty;

    /**
     * Initializes the NumMatrix with the given integer matrix.
     * Pre-computes the prefix sums in O(m * n) time.
     *
     * @param matrix The 2D input matrix.
     */
    public NumMatrix(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            this.isEmpty = true;
            this.prefixSum = null;
            return;
        }

        this.isEmpty = false;
        int rows = matrix.length;
        int cols = matrix[0].length;

        // Size is (rows + 1) x (cols + 1) to handle 0-index edge cases gracefully without "if" checks.
        this.prefixSum = new int[rows + 1][cols + 1];

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                // Formula: Current Cell + Area Above + Area Left - Overlap (Top-Left diagonal)
                this.prefixSum[r + 1][c + 1] = matrix[r][c] 
                                             + this.prefixSum[r][c + 1] 
                                             + this.prefixSum[r + 1][c] 
                                             - this.prefixSum[r][c];
            }
        }
    }

    /**
     * Calculates the sum of the elements inside the rectangle defined by its
     * upper-left corner (row1, col1) and lower-right corner (row2, col2).
     *
     * @param row1 The row index of the upper-left corner.
     * @param col1 The column index of the upper-left corner.
     * @param row2 The row index of the lower-right corner.
     * @param col2 The column index of the lower-right corner.
     * @return The sum of the region. Returns 0 if the matrix is empty.
     * @throws IllegalArgumentException if indices are out of bounds.
     */
    public int sumRegion(int row1, int col1, int row2, int col2) {
        if (this.isEmpty) {
            return 0;
        }
        
        // Optional: Input validation for robustness
        validateIndices(row1, col1, row2, col2);

        // Retrieve pre-computed sums
        // prefixSum[row2 + 1][col2 + 1] represents sum from (0,0) to (row2, col2)
        int totalArea   = this.prefixSum[row2 + 1][col2 + 1];
        int areaAbove   = this.prefixSum[row1][col2 + 1];
        int areaLeft    = this.prefixSum[row2 + 1][col1];
        int areaOverlap = this.prefixSum[row1][col1];

        // Apply Inclusion-Exclusion Principle
        return totalArea - areaAbove - areaLeft + areaOverlap;
    }

    private void validateIndices(int r1, int c1, int r2, int c2) {
        int R = prefixSum.length - 1;
        int C = prefixSum[0].length - 1;
        if (r1 < 0 || c1 < 0 || r2 >= R || c2 >= C || r1 > r2 || c1 > c2) {
             throw new IllegalArgumentException("Invalid sub-matrix indices.");
        }
    }
}